octree_web_viewer/src/backend.rs

```diff
@@ -1,45 +1,51 @@
+use actix_web::{
+    dev::Handler, http::ContentEncoding, AsyncResponder, Error, FromRequest, FutureResponse,
+    HttpRequest, HttpResponse, Json, Result,
+};
 use byteorder::{LittleEndian, WriteBytesExt};
 use cgmath::Matrix4;
-use iron;
-use iron::mime::Mime;
-use iron::prelude::*;
-use json;
+use futures::future::{result, Future};
 use point_viewer::octree::{self, Octree};
-use std::io::Read;
-use std::sync::{Arc, RwLock};
+use std::sync::Arc;
 use time;
-use urlencoded::UrlEncodedQuery;
 
 pub struct VisibleNodes {
-    octree: Arc<RwLock<Octree>>,
+    octree: Arc<dyn Octree>,
 }
 
 impl VisibleNodes {
-    pub fn new(octree: Arc<RwLock<Octree>>) -> Self {
+    pub fn new(octree: Arc<dyn Octree>) -> Self {
         VisibleNodes { octree }
     }
 }
 
-impl iron::Handler for VisibleNodes {
-    fn handle(&self, req: &mut Request) -> IronResult<Response> {
-        // TODO(hrapp): This should not crash on error, but return a valid http response.
-        let query = req.get_ref::<UrlEncodedQuery>().unwrap();
+// CAVEAT from actix docs
+//Be careful with synchronization primitives like Mutex or RwLock.
+//The actix-web framework handles requests asynchronously.
+//By blocking thread execution, all concurrent request handling processes would block.
+//If you need to share or update some state from multiple threads, consider using the actix actor system.
+
+impl<S> Handler<S> for VisibleNodes {
+    type Result = Result<HttpResponse, Error>;
+
+    fn handle(&self, req: &HttpRequest<S>) -> Self::Result {
         let matrix = {
             // Entries are column major.
-            let e: Vec<f32> = query.get("matrix").unwrap()[0]
+            let e: Vec<f32> = req
+                .query()
+                .get("matrix")
+                .expect("4x4 Matrix information expected")
```

Comment:
by replacing it I get a compiling error stating:
  the `?` operator can only be used in a function that returns `Result`
              or `Option` 

or

the trait `actix_web::ResponseError` is not implemented for `std::option::NoneError`
   |
   = note: required because of the requirements on the impl of `std::convert::From<std::option::NoneError>` for `actix_web::Error`

Reply:
error handling is now up for review